nikola/plugin_categories.py

```diff
@@ -72,45 +71,23 @@ def set_site(self, site):
         if not site.debug:
             self.logger.level = logging.INFO
 
+    def set_module_path(self, module_path):
+        """Set the plugin's module path."""
+        self.module_path = module_path
+
     def inject_templates(self):
         """Inject 'templates/<engine>' (if exists) very early in the theme chain."""
         try:
-            mod_candidate = None
-            # since https://github.com/tibonihoo/yapsy/pull/11 ,
-            # yapsy only adds each imported plugin to sys.modules
-            # under its modified, "unique" name (see early in
-            # PluginManager.loadPlugins), so we recreate the
-            # modified name here to find it. we fudge the serial
-            # number here, assuming that if a plugin is loaded
-            # under the same name multiple times, the location
-            # will also be the same, so we can just use 0.
-            possible_names = (
-                self.__class__.__module__,
-                "yapsy_loaded_plugin_" + self.__class__.__module__ + "_0",
-                "yapsy_loaded_plugin_" + self.name + "_0",
-            )
-            for possible_name in possible_names:
-                mod_candidate = sys.modules.get(possible_name)
-                if mod_candidate:
-                    break
-            if not mod_candidate:
-                # well, we tried. we wind up here for the dummy
-                # plugins; honestly I'm not sure exactly why/how,
-                # but they don't have templates, so it's okay
-                return
-            # Sorry, found no other way to get this
-            mod_path = mod_candidate.__file__
-            mod_dir = os.path.dirname(mod_path)
+            mod_dir = os.path.dirname(self.module_path)
             tmpl_dir = os.path.join(
                 mod_dir, 'templates', self.site.template_system.name
             )
             if os.path.isdir(tmpl_dir):
                 # Inject tmpl_dir low in the theme chain
                 self.site.template_system.inject_directory(tmpl_dir)
         except AttributeError:
-            # In some cases, __builtin__ becomes the module of a plugin.
-            # We couldn’t reproduce that, and really find the reason for this,
-            # so let’s just ignore it and be done with it.
+            # Just in case something went wrong and we don't have
```

Comment:
1. Does this ever happen with the base site + some plugins that need templates?
2. Please log a message (warning level?) to let users know if something explodes.

Reply:
Good question!

It doesn't happen at least on a test run on my blog (if I change the `pass` to `raise`). But it seems to me like there's no harm in being safe; if this ever *does* happen, it seems better to try accepting it and moving on than just blow up. I think in the worst case, we'll fail later when a required template isn't available (this is what happens in my 'real world' test when I get something wrong).

Adding a log message is a good idea, though - it doesn't seem like something that should *usually* happen, at least, so logging at warning level seems justified. I'll do that.